File: app/src/main/java/nl/rsdt/japp/service/NavigationLocationService.java
package nl.rsdt.japp.service;

import android.app.Service;
import android.content.ActivityNotFoundException;
import android.content.Intent;
import android.net.Uri;
import android.os.Binder;
import android.os.Handler;
import android.os.IBinder;
import android.os.Looper;
import android.widget.Toast;

import nl.rsdt.japp.R;
import nl.rsdt.japp.application.JappPreferences;
import nl.rsdt.japp.jotial.maps.NavigationLocationManager;

public class NavigationLocationService extends Service {
    Binder binder = new NavigationLocationBinder();


    public NavigationLocationService() {
        NavigationLocationManager locationManager = new NavigationLocationManager();
        locationManager.setCallback(new NavigationLocationManager.OnNewLocation() {
            @Override
            public void onNewLocation(nl.rsdt.japp.jotial.data.firebase.Location location) {
                if (JappPreferences.isNavigationPhone()) {
                    try {
                        String mesg = getString(R.string.location_received, location.createdBy, location.lat, location.lon);
                        showToast(mesg, Toast.LENGTH_SHORT);
                        switch (JappPreferences.navigationApp()) {
                            case GoogleMaps:
                                String uristr = getString(R.string.google_uri, Double.toString(location.lat),Double.toString(location.lon));
                                Uri gmmIntentUri = Uri.parse(uristr);
                                Intent mapIntent = new Intent(Intent.ACTION_VIEW, gmmIntentUri);
                                mapIntent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
                                mapIntent.setPackage("com.google.android.apps.maps");
                                startActivity(mapIntent);
                                break;
                            case Waze:
                                String uri = getString(R.string.waze_uri, Double.toString(location.lat),Double.toString(location.lon));
                                Intent wazeIntent =new Intent(Intent.ACTION_VIEW, Uri.parse(uri));
                                wazeIntent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
                                startActivity(wazeIntent);
                                break;
                            case OSMAnd:
                                String osmUri = getString(R.string.osmand_uri, Double.toString(location.lat),Double.toString(location.lon));
                                Intent osmIntent =new Intent(Intent.ACTION_VIEW, Uri.parse(osmUri));
                                osmIntent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
                                startActivity(osmIntent);
                                break;
                            case OSMAndWalk:
                                String osmuriwalk = getString(R.string.osmandwalk_uri, Double.toString(location.lat),Double.toString(location.lon));
                                Intent osmWalkIntent =new Intent(Intent.ACTION_VIEW, Uri.parse(osmuriwalk));
                                osmWalkIntent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
                                startActivity(osmWalkIntent);
                                break;
                            case Geo:
                                String geoUri = getString(R.string.geo_uri, Double.toString(location.lat),Double.toString(location.lon));
                                Intent geoIntent =new Intent(Intent.ACTION_VIEW, Uri.parse(geoUri));
                                geoIntent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
                                startActivity(geoIntent);
                                break;
                        }
                    } catch (ActivityNotFoundException e) {
                        System.out.println(e.toString());
                        String mesg = getString(R.string.navigation_app_not_installed, JappPreferences.navigationApp().toString());
                        showToast(mesg, Toast.LENGTH_SHORT);
                    }
                }
            }

            @Override
            public void onNotInCar() {
                String mesg = getString(R.string.fout_not_in_car);
                showToast(mesg, Toast.LENGTH_SHORT);
            }
        });
    }

    @Override
    public IBinder onBind(Intent intent) {
        // TODO: Return the communication channel to the service.
        return binder;
    }

    private void showToast(final String message, final int length){
        Handler handler = new Handler(Looper.getMainLooper());
        handler.post(new Runnable() {

            @Override
            public void run() {
                Toast.makeText(getApplicationContext(),
                        message,
                        length).show();
            }
        });
    }

    class NavigationLocationBinder extends Binder {
        public NavigationLocationService getInstance() {
            return NavigationLocationService.this;
        }
    }
}

File: app/src/main/java/nl/rsdt/japp/service/NavigationLocationService.kt
package nl.rsdt.japp.service

import android.app.Service
import android.content.ActivityNotFoundException
import android.content.Intent
import android.net.Uri
import android.os.Binder
import android.os.Handler
import android.os.IBinder
import android.os.Looper
import android.widget.Toast

import nl.rsdt.japp.R
import nl.rsdt.japp.application.JappPreferences
import nl.rsdt.japp.jotial.data.nav.Location
import nl.rsdt.japp.jotial.maps.NavigationLocationManager

class NavigationLocationService : Service(), NavigationLocationManager.OnNewLocation{
    private var binder: Binder = NavigationLocationBinder()

    init {
        val locationManager = NavigationLocationManager()
        locationManager.setCallback(this)
    }
    override fun onCreate(){
        AutoSocketHandler.init()
    }
    override fun onNewLocation(location: Location) {
        if (JappPreferences.isNavigationPhone) {
            try {
                val mesg = getString(R.string.location_received, location.username, location.lat, location.lon)
                showToast(mesg, Toast.LENGTH_SHORT)
                when (JappPreferences.navigationApp()) {
                    JappPreferences.NavigationApp.GoogleMaps -> {
                        val uristr = getString(R.string.google_uri, java.lang.Double.toString(location.lat), java.lang.Double.toString(location.lon))
                        val gmmIntentUri = Uri.parse(uristr)
                        val mapIntent = Intent(Intent.ACTION_VIEW, gmmIntentUri)
                        mapIntent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                        mapIntent.setPackage("com.google.android.apps.maps")
                        startActivity(mapIntent)
                    }
                    JappPreferences.NavigationApp.Waze -> {
                        val uri = getString(R.string.waze_uri, java.lang.Double.toString(location.lat), java.lang.Double.toString(location.lon))
                        val wazeIntent = Intent(Intent.ACTION_VIEW, Uri.parse(uri))
                        wazeIntent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                        startActivity(wazeIntent)
                    }
                    JappPreferences.NavigationApp.OSMAnd -> {
                        val osmUri = getString(R.string.osmand_uri, java.lang.Double.toString(location.lat), java.lang.Double.toString(location.lon))
                        val osmIntent = Intent(Intent.ACTION_VIEW, Uri.parse(osmUri))
                        osmIntent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                        startActivity(osmIntent)
                    }
                    JappPreferences.NavigationApp.OSMAndWalk -> {
                        val osmuriwalk = getString(R.string.osmandwalk_uri, java.lang.Double.toString(location.lat), java.lang.Double.toString(location.lon))
                        val osmWalkIntent = Intent(Intent.ACTION_VIEW, Uri.parse(osmuriwalk))
                        osmWalkIntent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                        startActivity(osmWalkIntent)
                    }
                    JappPreferences.NavigationApp.Geo -> {
                        val geoUri = getString(R.string.geo_uri, java.lang.Double.toString(location.lat), java.lang.Double.toString(location.lon))
                        val geoIntent = Intent(Intent.ACTION_VIEW, Uri.parse(geoUri))
                        geoIntent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                        startActivity(geoIntent)
                    }
                }
            } catch (e: ActivityNotFoundException) {
                println(e.toString())
                val mesg = getString(R.string.navigation_app_not_installed, JappPreferences.navigationApp().toString())
                showToast(mesg, Toast.LENGTH_SHORT)
            }
        }
    }
    override fun onNotInCar() {
        val mesg = getString(R.string.fout_not_in_car)
        showToast(mesg, Toast.LENGTH_SHORT)
    }

    override fun onBind(intent: Intent): IBinder? {
        // TODO: Return the communication channel to the service.
        return binder
    }

    private fun showToast(message: String, length: Int) {
        val handler = Handler(Looper.getMainLooper())
        handler.post {
            Toast.makeText(applicationContext,
                    message,
                    length).show()
        }
    }

    internal inner class NavigationLocationBinder : Binder() {
        val instance: NavigationLocationService
            get() = this@NavigationLocationService
    }
}
